rails_programming/turbo/stimulus_js.md

```diff
@@ -0,0 +1,287 @@
+
+An empty controller looks something like this:
+
+```javascript
+import { Controller } from "@hotwired/stimulus"
+export default class extends Controller {}
+```
+
+This controller does not do anything, it only shows how it imports tht base class `Controller` that we use a basis.
+As you see there is no mention of the name of the controller, the name is inferred from the file name 
+(`input_controller.js` becomes `input`).
+
+This creates a scope, all the following attributes will only work within the scope of that controller / dom element.
+
+### Triggering an action
+
+You learned how to trigger events with event listeners, in Stimulus instead you use `data-action` attributes to
+execute javascript to react to a user click or input.
+
+So instead of `document.querySelector("button).addEventListener("click", showAlert)` we write the following html
+
+```html
+<div data-controller="alert">
+  <button data-action="click->alert#show">Alert me!</button>
+</div>
+```
+
+Now clicking the button will trigger the action of the associated Stimulus controller:
+
+````js
+// app/javascript/controllers/alert_controller.js
+import { Controller } from "@hotwired/stimulus"
+export default class extends Controller {
+    show() {
+        alert("Hey from Stimulus");    
+    }
+}
+````
+
+### Selecting / targeting elements
+
+You learned how to access dom elements with selectors, such as `document.querySelector` and `document.getElementById`.
+Again Stimulus gives you a way to declare elements you want to select in the html:
+
+```html
+<div data-controller="greeter">
+  <input type="text" 
+         data-action="click->greeter#greet" 
+         data-greeter-target="name">
+    Alert me!
+  </input>
+  <div data-greeter-target="output"></div>
+</div>
+```
+
+Notice the `data-greeter-target`. Targets can then be used in your controller:
+
+````js
+// app/javascript/controllers/greeter_controller.js
+import { Controller } from "@hotwired/stimulus"
+export default class extends Controller {
+    static targets = ["name", "output"]
+    greet() {
+        this.nameTarget.html = this.inputTarget.value     
+    }
+}
+````
+
+You need to declare your targets, but once you do, you can get the dom-element by using `this.nameTarget`, or 
+if you have multiple: `this.nameTargets`. If you need to make your controllers to be smart about targets, you can
+also ask whether certain targets are available in the html with `this.hasNameTarget`.
+
+### Keeping state
+
+A Stimulus controller can have its internal state, meaning you can keep data in the controller. Though usually you want
+to keep this to a minimum.
+
+So anything you define on `this` is available throughout the controller:
+```js
+export default class extends Controller {
+    connect() {
+        this.count = 0
+    }
+  
+    addOne() {
+        this.count++
+    }
+}
+```
+
+Stimulus also lets you declare specific value attributes, that allow you to listen to changes:
+
+````js
```

Comment:
Is there a reason for four backticks? I was just wondering...

Reply:
Nope, no reason just hitting the button to many times :-) Fixed that throughout the document